keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -0,0 +1,372 @@
+Both are rather usage based than "number of protocols" based, though it is true that in case of NLB TCP, UDP and TLS has separated quotas in the pricing units.
+
+A user can ask for an internal Load Balancer via a K8s Service definition that also has the annotation `service.beta.kubernetes.io/aws-load-balancer-internal: 0.0.0.0/0`. So far the author could not find any difference in the usage and pricing of those when compared to the external LBs - except the pre-requisite of a private subnet on which the LB can be deployed.
+
+Summary: AWS CPI is the current bottleneck with its TCP-only limitation. As long as it is there the implementation of this feature will have no effect on the AWS bills.
+
+#### Azure
+
+Azure CPI LB documentation: https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/docs/services/README.md
+
+The Azure CPI already supports the usage of both UDP and TCP protocols in the same Service definition. It is achieved with the CPI specific annotation `service.beta.kubernetes.io/azure-load-balancer-mixed-protocols`. If this key has value `true` in the Service definition the Azure CPI adds the other protocol value (UDP or TCP) to its internal Service representation, and as a consequence it also manages twice the amount of load balancer rules for the specific frontend. 
+
+Only TCP and UDP are supported in the current mixed protocol configuration.
+
+The Azure Load Balancer supports only TCP and UDP as protocols. HTTP support would require the usage of the Azure Application Gateway. I.e. HTTP and L4 protocols cannot be used on the same LB instance/IP address
+
+Basic Azure Load Balancers are free.
+
+Pricing of the Standard Azure Load Balancer is based on load-balancing rules and outbound rules.  There is a flat price up to 5 rules, on top of which every new forwarding rule has additional cost. 
+https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-overview#pricing
+
+A user can ask for an internal Azure Load Balancer via a K8s Service definition that also has the annotation `service.beta.kubernetes.io/azure-load-balancer-internal: true`. There is not any limitation on the usage of mixed service protocols with internal LBs in the Azure CPI.
+
+Summary: Azure already has mixed protocol support for TCP and UDP, and it already affects the bills of the users. The implementation of this feature may require some work in Azure CPI.
+
+#### GCE
+
+The GCE CPI supports both TCP and UDP protocols in Services. Other protocols are not supported.
+
+GCE/GKE creates Network Load Balancers based on the K8s Services with type=LoadBalancer. In GCE there are "forwarding rules" that define how the incoming traffic shall be forwarded to the compute instances. A single forwarding rule can support either TCP or UDP but not both. In order to have both TCP and UDP forwarding rules we have to create separate forwarding rule instances for those. Two or more forwarding rules can share the same external IP if
```

Comment:
I think the issue/PR has a comment with someone mentioning that this worked for them. We should crosslink comment of the steps that they took as a demonstration of this working.

Reply:
Fixed. I added a link here to the comment.